docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+It is derived from a [WebAssembly Interface Types](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md)
+file, maintained by hand by the package author.
+
+This descriptor is used for two purposes:
+
+1. To generate language-specific bindings to the interface(s) described. WIT is an Interface Description Language (IDL),
+describing interfaces in a language-agnostic way, but defined in terms of a [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md),
+allowing efficient cross-language interop. We follow the semantics of the Wasm canonical ABI, but layered on top of
+our canonical Miden ABI, which describes how the Wasm abstract machine is translated to the Miden abstract machine.
+
+2. To provide the compiler with the necessary type information needed to emit lifting/lowering code at call boundaries,
+as well as mapping imports (from the packages' dependencies) to their corresponding MAST roots, so that calls use those
+roots rather than the fully-qualified names. This ensures that the code that was compiled and published is always the
+exact code that gets executed, in perpetuity.
+
+Because the type descriptor is in binary form, to view it in the more human-friendly WIT format, you
+will need to use [wasm-tools](https://github.com/bytecodealliance/wasm-tools), like so:
+
+    wasm-tools component wit extracted-package/component.wasm
+
+The contents of `component.wasm` here is not an actual Wasm module, but simply the binary representation
+of the WIT package.
+
+## Directory Layout
+
+The canonical layout of a package when extracted as a directory, is as follows:
+
+    name@version/
+    |- manifest.toml
+    |- component.wasm
+    |- lib/
+       |- 0x00000000.mast
+       |- 0x00000001.mast
+    |- rodata/
+       |- 0x0.bin
+    |- etc/
+       |- component.sourcemap
+       |- docs/
+          |- index.html
+       |- vendor/
+          |- foo@0.1.0.mpkg
+
+### Manifest
+
+The `manifest.toml` file contains all of the top-level metadata about the package. When produced as a result
+of extracting a package, it's contents may differ from the `manifest.toml` used to create the package. This
+is because we provide some helpful functionality in the manifest for creating packages by hand, that are
+translated into canonical form when the package is created. These "canonicalizations" are covered below
+in each relevant section.
+
+> [!NOTE]
+> TOML was chosen as the format for the manifest due to it being a nicer format for
+> writing by hand, while being strongly typed, unambiguous, and supported by good
+> tooling in most languages.
+
+The manifest begins with basic metadata about the package:
+
+```toml
+[package]
+name = "my-package"
+version = "0.1.0"
+miden-version = "0.8.0"  # Optional
```

Comment:
We certainly could - I was thinking of it a bit like `rust-version`, where you can signal that a given package cannot be compiled/used by versions of the tooling earlier than a given version due to the use of some specific feature or implementation detail. That won't be the case for all packages, and doesn't actually matter if it is missing or incorrect, since ultimately what will happen is that the code will fail at runtime rather than us being able to raise an error at compile-time.

So I think it largely makes more sense for it to be optional, but I could see a case being made for it to be required.

Reply:
As mentioned in one of the comments above, maybe this should go into the `.mast` file format.